### PR TITLE
Adding DS3231 component support

### PR DESCRIPTION
**Pull request by KerryRJ**
I have added a Ninja flash task to VS2019 but have not yet figured out how to automatically obtain the COM port used by the ESP32. So this has to be manually edited.

I would like to know if I should be creating a self contained example in examples/esp/ds3231 folder rather than working on adding a ds3231 test. This would be so that anyone could just copy the sample and run it. Thoughts?

### CODE REVIEW

**Comment by PerMalmberg** [comment]
I'd much prefer a test project in test/hw_ds3231 as that'll follow the same structure as the other test/example projects which all are easily runable by editing the top CMakeLists.txt. The test may be more than "just a test" if you so wish.

File: lib/smooth/application/io/i2c/DS3231.cpp [hunk 1]
```diff
@@ -0,0 +1,658 @@
+// Smooth - C++ framework for writing applications based on Espressif's ESP-IDF.
```
Comment (PerMalmberg) [inline lib/smooth/application/io/i2c/DS3231.cpp]: As the license is changing, please update the header to use the Apache 2.0 notice as outlined in the new CONTRIBUTING.md

File: lib/smooth/application/io/i2c/DS3231.cpp [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+
+                std::vector<uint8_t> data;
+                data.push_back(Register::Alarm1);
+                data.push_back(rate >= SecondsMatch ? Uint8_t2BCD(static_cast<uint8_t>(time.seconds().count())) : 0x80);
+                data.push_back(rate >= MinutesAndSecondsMatch ? Uint8_t2BCD(
+                    static_cast<uint8_t>(time.minutes().count())) : 0x80);
+                data.push_back(rate >=
+                HoursMinutesAndSecondsMatch ? Uint8_t2BCD(static_cast<uint8_t>(time.hours().count())) : 0x80);
+                data.push_back(
+                    rate >= DateHoursMinutesAndSecondsMatch
```
Comment (PerMalmberg) [inline lib/smooth/application/io/i2c/DS3231.cpp]: Can we avoid nested ?: operators? It makes it somewhat hard to follow, especially with the long statements.

File: lib/smooth/application/io/i2c/DS3231.cpp [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+                auto time = make_time(timePoint - timePointWithDaysResolution);
+
+                std::vector<uint8_t> data;
+                data.push_back(Register::Alarm2);
+                data.push_back(rate >= MinutesMatch ? Uint8_t2BCD(static_cast<uint8_t>(time.minutes().count())) : 0x80);
+                data.push_back(rate >= HoursAndMinutesMatch ? Uint8_t2BCD(
+                    static_cast<uint8_t>(time.hours().count())) : 0x80);
+                data.push_back(
+                    rate >= DateHoursAndMinutesMatch
+                    ? Uint8_t2BCD(static_cast<uint8_t>(unsigned(date.day())))
```
Comment (PerMalmberg) [inline lib/smooth/application/io/i2c/DS3231.cpp]: Same as before - nested ?: operators.

**Comment by PerMalmberg** [comment]
Btw, I'd love to hear what you think about the I2C abstraction layer in Smooth (and thoughts on other parts too)

File: lib/compiler_options.cmake [hunk 4]
```diff
@@ -15,13 +15,16 @@
 # along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 function(set_compile_options target)
+	# not sure why but I can only get intellisense in VS2019 working with -std=c++17
```
[the remark this answers is not in the document]
Comment (KerryRJ) [reply lib/compiler_options.cmake]: The if(MSVC) test doesn't help as we're using Cmake with the ESP-IDF tool set so the compiler is GNU 8.2.0 and not MSVC.

Removing this as it appears to be fixed in VS2019 16.3 Preview 3.

File: CMakeSettings.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
+﻿{
```
[the remark this answers is not in the document]
Comment (KerryRJ) [reply CMakeSettings.json]: These are VS2019 ESP-IDF cross compile settings.

I forgot to add them in initially when first trying to compile smooth so I thought I'd add it in to make it easier for myself and others, in future.

Is it better to have this in a document file.

File: lib/smooth/include/smooth/application/io/i2c/DS3231.h [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+            /// Forces the temperature sensor to convert the temperature into digital code,
+            /// and execute the TCXO algorithm to update the capacitance array to the oscillator.
+            /// \return true if the 32kHz Output was enabled,
+            ///		false if failure writing the bit
+            ///
+            /// Setting this bit to 1 forces the temperature sensor to convert the temperature
+            /// into digital code and execute the TCXO algorithm to update the capacitance array to the oscillator.
+            /// This can only happen when a conversion is not already in progress.
+            /// The user should check the status bit BSY before forcing the controller to start a new TCXO execution.
+            /// A user - initiated temperature conversion does not affect the internal 64 - second update cycle.
```
[the remark this answers is not in the document]
Comment (KerryRJ) [reply lib/smooth/include/smooth/application/io/i2c/DS3231.h]: Uncrustify is not adding the spaces. It may have been when I pasted the text in from the document.

File: sdkconfig [hunk 7]
```diff
@@ -66,11 +66,11 @@ CONFIG_ESPTOOLPY_MONITOR_BAUD_115200B=y
 # CONFIG_ESPTOOLPY_MONITOR_BAUD_OTHER is not set
```
[the remark this answers is not in the document]
Comment (KerryRJ) [reply sdkconfig]: There was something with the defaults that I had to change to get the Wemos Lolin32 that I was testing on to not panic. I actually think I had to turn off the PSRAM as it does not have any.
I'll confirm this and if it is that setting do I upload it or not?

File: lib/smooth/include/smooth/application/io/i2c/DS3231.h [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+                             system_clock::time_point const& timePoint = system_clock::from_time_t(0));
+
+            //AlarmTwo GetAlarmTwo();
+
+            /// Sets alarm two
+            /// \param timePoint The date and time of the alarm
+            /// \param rate The rate of the alarm
+            /// \return true if the alarm was set,
+            ///		false if there was a failure writing the alarm data
+            bool SetAlarmTwo(AlarmTwoRate rate = OncePerMinute,
```
[the remark this answers is not in the document]
Comment (KerryRJ) [reply lib/smooth/include/smooth/application/io/i2c/DS3231.h]: See alarm rate One comment above except that the default alarm rate is OncePerMinute.

File: lib/smooth/include/smooth/application/io/i2c/DS3231.h [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+            bool SetDateTime(system_clock::time_point& dateTime);
+
+            //AlarmOne GetAlarmOne();
+
+            /// Sets alarm one
+            /// \param timePoint The date and time of the alarm
+            /// \param rate The rate of the alarm
+            /// \return true if the alarm was set,
+            ///		false if there was a failure writing the alarm data
+            bool SetAlarmOne(AlarmOneRate rate = OncePerSecond,
```
[the remark this answers is not in the document]
Comment (KerryRJ) [reply lib/smooth/include/smooth/application/io/i2c/DS3231.h]: The time point is not required if the default alarm rate is OncePerSecond. Any alarm set in the past will theoretically not fire the interrupt pin.

I thought that the OncePerSecond may well be the most used option for this alarm rate.

I'm not sure what to set the time point to as a default.

File: lib/smooth/include/smooth/application/io/i2c/DS3231.h [hunk 10]
```diff
@@ -0,0 +1,351 @@
+// Smooth - C++ framework for writing applications based on Espressif's ESP-IDF.
+
```
[the remark this answers is not in the document]
Comment (KerryRJ) [reply lib/smooth/include/smooth/application/io/i2c/DS3231.h]: What do I need to look for to make sure this is in UTF8?
I thought it would be by default. Is it possible that my Git for Windows is messing with file formats?

File: lib/smooth/include/smooth/application/io/i2c/DS3231.h [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+            enum AlarmTwoRate
+            {
+                OncePerMinute = 0,
+                MinutesMatch,
+                HoursAndMinutesMatch,
+                DateHoursAndMinutesMatch,
+                DayHoursAndMinutesMatch
+            };
+
+            //void event(const smooth::core::io::InterruptInputEvent& value) override;
```
[the remark this answers is not in the document]
Comment (KerryRJ) [reply lib/smooth/include/smooth/application/io/i2c/DS3231.h]: This is a point that I wanted to raise. I tried to get this device to accept a task so that I could fire an event if AlarmOne was triggered and another event if AlarmTwo was triggered. I got to a point where I needed a task to be passed through when the I2C device is created but could not find a solution due to my limited C++ skills.

Do you have a sample for this?
I think in the general this option would be the way the device would mostly be used but if the device is used to wake the ESP32 from deep sleep then there may need to be an option to turn off the handling of the interrupt and related events. Otherwise it may just be better to continue as is and have the user code either of the functionalities. What do you think?

File: lib/compiler_options.cmake (hunk 4 above)
[the remark this answers is not in the document]
Comment (PerMalmberg) [reply lib/compiler_options.cmake]: Hm, did you write that this was fixed in an update of VS?

Anyway, since ESP-IDF doesn't use MSYs any more, I think the way to wrap this using [CMAKE_HOST_SYSTEM_NAME](https://cmake.org/cmake/help/latest/variable/CMAKE_HOST_SYSTEM_NAME.html#variable:CMAKE_HOST_SYSTEM_NAME)

File: lib/smooth/include/smooth/application/io/i2c/DS3231.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (PerMalmberg) [reply lib/smooth/include/smooth/application/io/i2c/DS3231.h]: I don't know how it looks in VS, but in VS Code, there's a text you can click in in the lower right corner:
image

File: lib/smooth/include/smooth/application/io/i2c/DS3231.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (PerMalmberg) [reply lib/smooth/include/smooth/application/io/i2c/DS3231.h]: I'm guessing you had a look at the [hw_interrupt_queue](https://github.com/PerMalmberg/Smooth/tree/master/test/hw_interrupt_queue) example?

The main App is a task, which is passed as the [second argument](https://github.com/PerMalmberg/Smooth/blob/master/test/hw_interrupt_queue/hw_interrupt_queue.cpp#L27) at this line. Likewise you could pass any core::Task to InterruptQueue. The reason a Task is passed is to tell in what context the event shall be run, i.e. what task actually calls  `event(const smooth::core::io::InterruptInputEvent& value)`.

As for turning things on and off, the `core::io::InterruptInput` doesn't currently support destruction, i.e. once created it is expected to remain alive until restart of the device. That said, I do believe it would be as simple as adding a destructor and calling `gpio_isr_handler_remove(io);` Once that is in place, you could probably place an `InterruptInput` in a `std::unique_ptr` and create/destory it on the fly.

```c++
InterruptInput::~InterruptInput()
{
    gpio_isr_handler_remove(io);
}
```

Since class members are guaranteed to be valid while the destructor is running, the entire object is intact, even if the interrupt happens while the destructor is running. This doesn't hold true for derived classes so I think it is a good idea that `IntteruptInput` is made a [final](https://en.cppreference.com/w/cpp/language/final) class to prevent its use in inheritance.

File: lib/smooth/include/smooth/application/io/i2c/DS3231.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (PerMalmberg) [reply lib/smooth/include/smooth/application/io/i2c/DS3231.h]: > I'm not sure what to set the time point to as a default.

None, let the caller of the method provide the time. IMHO, `OncePerSecond` is the odd one here and adjusting method signatures to fit that makes the API less clear.  You could have a dedicated method for setting the alarm to `OncePerSecond`.

File: sdkconfig (hunk 7 above)
[the remark this answers is not in the document]
Comment (PerMalmberg) [reply sdkconfig]: That's up to the user of the library - they are already advised that they should adjust their sdkconfig to their use case in the README.md. Don't upload.

File: CMakeSettings.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (PerMalmberg) [reply CMakeSettings.json]: Ok, feel free to add a section to the README.md dedicated to VS, unless Espressif already has this in their documentation in which case a link would be enough?

**Comment by PerMalmberg** [comment]
@KerryRJ just checking if I have missed a message from you regarding the requested changes in this PR as no changes has been made in 10 days?

**Comment by KerryRJ** [comment]
Nope.

I am still busy working on the changes.

Kind Regards

Kerry Jones
Owning Member

Info-Overdrive cc
[IO Drive square logo - blue - small]

m. +27(0)612403717
e. kerry@iodrive.co.za<mailto:kerry@iodrive.co.za>
a. F3 Chief Mhlabunzima Rd., Pietermaritzburg, 3201

Website<http://www.iodrive.co.za/>        Email<mailto:kerry@iodrive.co.za>            IM<sip:kerry@iodrive.co.za>

**Comment by PerMalmberg** [comment]
@KerryRJ I've added a first version of a CI pipeline. It currently lists the status as waiting and I think it'll start a build once you hit that "Update branch" button, alternatively on your next push.

**Comment by PerMalmberg** [comment]
@KerryRJ,

FYI: There are two major changes in master now:
- Removal of names for queues.
- New log framework, {fmt}, https://fmt.dev

**Comment by PerMalmberg** [comment]
@KerryRJ Have you abandoned this PR?

**Comment by KerryRJ** [comment]
Close it for now.
